### PR TITLE
Refactoring Part 3 - CLuaFunctionParser

### PR DESCRIPTION
**Pull request by TracerDS**
#3493 related

There wont be a merge conflict as this PR is an extension to #3493

### CODE REVIEW

File: Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h [hunk 1]
```diff
@@ -723,30 +725,34 @@ struct CLuaFunctionParser<ErrorOnFailure, ReturnOnFailure, Func> : CLuaFunctionP
...
-            }
-            else
-            {
-                pScriptDebugging->LogCustom(L, strError.c_str());
-                lua::Push(L, ReturnOnFailure);
-            }
+            strError = e.what();
+            bIsCustomError = true;
+        }
+        if (!strError.empty())
```
Comment (tederis) [inline Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h]: Should be `if (strError.empty())`

File: Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h [hunk 2]
```diff
@@ -19,6 +19,15 @@ class CLuaArgument;
 #include "lua/LuaBasic.h"
 #include <lua/CLuaMultiReturn.h>
 
+class ArgumentParserError
```
Comment (tederis) [inline Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h]: I would name it `LuaFunctionError` and add a new field `m_bWarning`. So you can inherit a class from it and set whether it should produce error or warning.

Comment (TracerDS) [reply Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h]: Good idea, thanks ❤

File: Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h [hunk 3]
```diff
@@ -192,79 +205,79 @@ struct CLuaFunctionParserBase
         // primitive types
         if constexpr (std::is_same_v<T, std::string> || std::is_same_v<T, std::string_view>)
             return (iArgument == LUA_TSTRING || iArgument == LUA_TNUMBER);
-        if constexpr (std::is_same_v<T, bool>)
+        else if constexpr (std::is_same_v<T, bool>)
```
Comment (botder) [inline Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h:208]: There is no need to add an `else` here and and in the changed lines below. The _if statement_ above always ends in a _return_.

Comment (TracerDS) [reply Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h:208]: > There is no need to add an `else` here and below. The _if statement_ above always ends in a _return_.

I added `else`s due to the nature of `if constexpr`. Not putting else in `if constexpr` will not disregard the latter statements. In other function it ends up with a compile time error.
C++ is not evolved yet for it to recognize early returns in `if constexpr` statements,

Comment (botder) [reply Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h:208]: Good to know, I expected the compiler to be better.

**Review by botder — APPROVED** [review]

**Pull request merged**